Add lockstep

julianoes opened the pull request:
These are the changes required to get lockstep running with PX4.
https://github.com/PX4/Firmware/pull/10648

This adds the option for lockstep simulation. Lockstep is off be default
unless `enable_lockstep` is set to true in the vehicle's SDF file.

Basically, lockstep means that PX4 is "stepped" forward by each
hil_sensor MAVLink message containing a timestamp which is then the time
set for PX4. Then, this plugin waits for the actuator_control message
coming back from PX4. Once that has arrived, we move on which means we
resume the simulation to get a new IMU sample.

Therefore there are two locations where we wait:
1. In `OnUpdate` we wait for the next IMU sample in case the callback
   for the IMU message subscription has not triggered yet.
2. Then later in `OnUpdate` we wait for the actuator control message to
   come back from PX4. Basically we use `poll` and `recv` until we have
   parsed the message that we're looking for.

Note that we "freewheel", so we send hil_sensor messages without waiting
for actuator_control messages for initialization until PX4 starts
sending actuator_control messages. Once we have received the first
control message, we start the "lockstepping".

julianoes committed:
msgs: add sequence and timestamp to IMU message

We need this for lockstep where we don't want to miss an Imu sample.

julianoes committed:
gazebo_imu_plugin: set seq and time in IMU message

julianoes committed:
gazebo_imu_plugin: remove newlines and comment

julianoes committed:
gazebo_imu_plugin: use queue of 10 for IMU msg pub

The problem that we're trying to work around here is that sometimes the
mavlink plugin misses an IMU sample especially if the CPU load is high.

What we can do is wait in `OnUpdate` in the mavlink plugin until we have
received the callback from the publication here. However, if we're
unlucky we won't even get to `OnUpdate` in the mavlink plugin before the
IMU plugin publishes yet another sample. In that case we overwrite the
sample if we don't have a queue > 1 which is what is changed in this
commit.

julianoes committed:
gazebo_mavlink_interface: rename variable name

I don't know what was meant with first referenc, really what it probably
means is first actuator message.

julianoes committed:
gazebo_mavlink_interface: miscellaneous tiny fixes

julianoes committed:
gazebo_mavlink_interface: added TCP support

This adds support to use TCP instead of UDP to connect to PX4 SITL.
By default TCP is off unless `use_tcp` is set to true in the vehicle's
SDF file.

julianoes committed:
gazebo_mavlink_interface: add lockstep

This adds the option for lockstep simulation. Lockstep is off be default
unless `enable_lockstep` is set to true in the vehicle's SDF file.

Basically, lockstep means that PX4 is "stepped" forward by each
hil_sensor MAVLink message containing a timestamp which is then the time
set for PX4. Then, this plugin waits for the actuator_control message
coming back from PX4. Once that has arrived, we move on which means we
resume the simulation to get a new IMU sample.

Therefore there are two locations where we wait:
1. In `OnUpdate` we wait for the next IMU sample in case the callback
   for the IMU message subscription has not triggered yet.
2. Then later in `OnUpdate` we wait for the actuator control message to
   come back from PX4. Basically we use `poll` and `recv` until we have
   parsed the message that we're looking for.

Note that we "freewheel", so we send hil_sensor messages without waiting
for actuator_control messages for initialization until PX4 starts
sending actuator_control messages. Once we have received the first
control message, we start the "lockstepping".

julianoes committed:
gazebo_mavlink_interface: just set wrong params

Instead of crashing when timing params are set incorrectly, we can just
set them. This way we don't need to change all world files just yet and
stay backwards compatible.

For now lockstep is limited to just 250 Hz with 4ms step size.

julianoes committed:
models: enable lockstep and TCP everywhere

Let's try this :).

julianoes committed:
worlds: for now we need to simulate at 250 Hz

Lockstep currently only works with 250 Hz, so we change it for now. If
we find that the simulation fidelity is not enough, we can improve it
and increase the rate later.

julianoes committed:
gazebo_mavlink_interface: abort if world incorrect

For now lockstep SITL only works at 250 Hz, so therefore the
real_time_update_rate needs to be set to 250 in all used world files and
the max_step_size needs to be 0.004.

This can be improved later if we decide we need more fidelity.

julianoes committed:
gazebo_mavlink_interface: spaces, not tabs!

julianoes committed:
gazebo_mavlink_interface: ifdefs for Gazebo <= 7

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -181,12 +180,78 @@ void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf
+    presetManager->GetCurrentProfileParam("real_time_update_rate", param);
+    double real_time_update_rate = boost::any_cast<double>(param);
+    const double correct_real_time_update_rate = 250.0;
+    if (real_time_update_rate != correct_real_time_update_rate)
```
Since we update `real_time_update_rate` below, do we need to ensure it's set to 250?

julianoes replied:
Yes, I realized it actually doesn't work if I change it later. It must get initialized earlier and then not updated.

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -367,18 +478,29 @@ void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf
+
+  if (previous_imu_seq_ > 0) {
+    while (previous_imu_seq_ == last_imu_message_.seq() && IsRunning()) {
+      std::this_thread::sleep_for(std::chrono::microseconds(1));
```
Can we do this event-based with a condition variable? Also is this thread-safe?

julianoes replied:
I suppose we can. Thread-safe? Not so much. I'll have to add a lock.

julianoes replied:
Or a `std::future`/`std::promise`?

TSC21 replied:
The `std::future` seems like a good approach.

julianoes replied:
I've tried this in 0a7b8769b3548b5f37db1fafde6ae4fe7e4db582.

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -559,79 +704,66 @@ void GazeboMavlinkInterface::ImuCallback(ImuPtr& imu_message) {
+  hil_state_quat.yacc = accel_true_b.Y() * 1000;
+  hil_state_quat.zacc = accel_true_b.Z() * 1000;
 
+  if (!hil_mode_ || (hil_mode_ && !hil_state_level_)) {
```
The logic here was
```
if (hil_mode_) {
      if (hil_state_level_){
```
Was this wrong before?

julianoes replied:
Good catch, let me check.

julianoes replied:
You're right, fixed in 1d910009a2694a6aa68c2dc232f850fa84473b00.

julianoes committed:
gazebo_mavlink_interface: fix if logic regression

This was incorrectly changed previously, and spotted in a review by
@bkueng.

TSC21 commented on models/matrice_100/matrice_100.sdf.jinja:
```diff
@@ -293,6 +293,8 @@
       <opticalFlowSubTopic>/px4flow/link/opticalFlow</opticalFlowSubTopic>
       <lidarSubTopic>/sf10a/link/lidar</lidarSubTopic>
+      <enable_lockstep>true</enable_lockstep>
+      <use_tcp>true</use_tcp>
```
Do we want TCP and lockstep to be true by default?

julianoes replied:
I'd say lockstep yes. TCP can be discussed, I have no strong opinion and @bkueng is right that it's not necessarily needed because we wait anyway for messages from the other side, so buffers won't fill up and we won't have any drops as long as everything is on localhost. Once we use SITL over the network, we should use TCP.

julianoes committed:
gazebo_mavlink_interface: use promise no usleep

Instead of sleeping until the ImuCallback has happened, it's probably
better to use a future/promise.

julianoes committed:
gazebo_mavlink_interface: add missing include

TSC21 reviewed (approved):
LGTM!

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -363,22 +477,40 @@ void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf
+  }
+
+  // Reset the future/promise. We need to do this no matter what, otherwise
+  // we rist that the promise is set twice in a row on startup.
```
typo: `risk`
And the way I see this you have a race condition here: you reset the promise, but the imu callback could already have been called another time (or does gazebo prevent this in any way?).
I find it strange that this callback runs on a different thread from the IMU callback.

julianoes replied:
Hm, right. So we have to add a mutex as well?

julianoes replied:
I see this in some of the tests:
```
terminate called after throwing an instance of 'std::future_error'
  what():  std::future_error: Promise already satisfied
Aborted (core dumped)
```
So I might just have to re-implement this using a condition variable.

bkueng replied:
Yes I think it's better suited for that use-case.
